:bug: - #234 - fix unrelated zaaktypen being shown in select on vario…

svenvandescheur opened the pull request:
…us screens

closes #234

SilviaAmAm commented on backend/src/openarchiefbeheer/zaken/utils.py:
```diff
@@ -20,6 +20,7 @@
 
 from openarchiefbeheer.utils.results_store import ResultStore
 
+from ..destruction.constants import ListItemStatus
```
```suggestion
from openarchiefbeheer.destruction.constants import ListItemStatus
```

SilviaAmAm commented on backend/src/openarchiefbeheer/zaken/utils.py:
```diff
@@ -76,7 +77,14 @@ def expand_procestype(zaak: dict) -> dict:
-def retrieve_zaaktypen_choices() -> list[DropDownChoice]:
+if TYPE_CHECKING:
+    # Due to circular import
+    from ..destruction.models import DestructionList
```
```suggestion
    from openarchiefbeheer.destruction.models import DestructionList
```

SilviaAmAm commented on backend/src/openarchiefbeheer/zaken/utils.py:
```diff
@@ -95,11 +103,18 @@ def retrieve_zaaktypen_choices() -> list[DropDownChoice]:
-        Zaak.objects.all()
-        .values_list("_expand__zaaktype__identificatie", flat=True)
-        .distinct()
+    zaak_qs = (
+        Zaak.objects.filter(
+            url__in=destruction_list.items.filter(
+                status=ListItemStatus.suggested  # TODO: Expand?
+            ).values_list("zaak", flat=True)
+        )
+        if destruction_list
+        else Zaak.objects.all()
```
Maybe more legible?
```python
zaak_qs = Zaak.objects.all()
if destruction_list:
    zaak_qs = zaak_qs.filter(
            url__in=destruction_list.items.filter(
                status=ListItemStatus.suggested  # TODO: Expand?
            ).values_list("zaak", flat=True)
    )
```
What do you mean with the `#TODO expand`?

SilviaAmAm commented on backend/src/openarchiefbeheer/destruction/api/filtersets.py:
```diff
@@ -76,10 +78,28 @@ def filter_destruction_list_uuid(
 
 
 class DestructionListReviewItemFilterset(FilterSet):
```
Would this not only solve the problem on the 'changes requested' view, but not on the 'normal' detail view of a destruction list?

SilviaAmAm commented on backend/src/openarchiefbeheer/zaken/api/serializers.py:
```diff
@@ -86,3 +86,21 @@ class SelectielijstklasseChoicesQueryParamSerializer(serializers.Serializer):
+        required=False,
+    )
+
+    def to_internal_value(self, data: dict) -> dict:
+        from openarchiefbeheer.destruction.models import DestructionList
+
+        internal_value = super().to_internal_value(data)
+
+        if destruction_list := internal_value.get("destruction_list"):
+            internal_value["destruction_list"] = DestructionList.objects.get(
+                uuid=destruction_list
+            )
+        return internal_value
```
This feels a bit like a hack? :thinking: 

Maybe a `validate` that checks that the destruction list with that UUID exists, but then leave getting the destruction list to the view?

SilviaAmAm reviewed (commented):
I only had a quick look at the backend, will give a more thorough review tomorrow!

SilviaAmAm commented on backend/src/openarchiefbeheer/destruction/api/filtersets.py:
I am thinking: to get around the problem that we also need to filter the cases on the page where the record manager processes the requested changes, maybe the `ZaaktypenChoicesView` could take also a `review` param to filter the cases? :thinking: Then we might not need to have code in multiple places to handle filtering the zaken :thinking:

Commit:
:sparkles: [#234] Get zaaktype choices for destruction_list

Commit:
:white_check_mark: [#234] Test retrieving zaaktype choices for list

Commit:
:construction: [#234] Fix frontend

Co-authored-by: svenvandescheur <svenvandescheur@gmail.com>

Commit:
:construction: [#234] Filter zaaktypechoices on review id

Commit:
:sparkles: [#234] Add filtering on review

Commit:
:white_check_mark: [#234] Test filtering on review

svenvandescheur commented on frontend/src/lib/api/private.ts:
```diff
@@ -11,6 +10,11 @@ export type ZaaktypeChoice = Option & {
   extra: string;
 };
 
+export type ZaaktypeChoicesQueryParam = {
```
Suggestion:

Remove `ZaaktypeChoicesQueryParam` (see next)

Motivation:

See next

svenvandescheur commented on frontend/src/pages/destructionlist/hooks/useDataGridProps.tsx:
```diff
@@ -48,6 +52,7 @@ export function useDataGridProps(
   paginatedResults: PaginatedDestructionListItems | PaginatedZaken,
   selectedResults: (Zaak | { url: string })[],
   actions?: DataGridAction[],
+  zaaktypeChoicesQueryParams?: ZaaktypeChoicesQueryParam,
```
Suggestion:

Replace `zaaktypeChoicesQueryParams` with two separate parameters:

```
destructionList?: DestructionList;
review?: Review
```

And have this hook figure out what to use.

Motivation:

This may keep things cleaner in the future duplication when some other functionality needs to access values from the destruction list and / or review.

svenvandescheur commented on frontend/src/lib/api/private.ts:
```diff
@@ -35,10 +39,11 @@ export async function listSelectieLijstKlasseChoices(
-    const promise: Promise<ZaaktypeChoice[]> = response.json();
-    return promise;
-  });
+export async function listZaaktypeChoices(
```
Suggestion:

Use function arguments for `listReviewItems`.

```
listZaaktypeChoices(destructionListUuid, reviewPk)
```

Motivation:

I'd like to have a more standardizes interface for these API accessors. The main identifiers are typically a function argument and other parameters use a custom params type.

I think the function arguments are cleaner here.

P.S.:

Is there a reasons we identify a list by uuid and a review by pk?

SilviaAmAm replied:
No, at some point we wanted to move to UUIDs, but we only go to migrating the destruction list and not all other resources :S

svenvandescheur reviewed (commented):
Some suggestions for the frontend bit.

Commit:
:ok_hand: [#234] PR Feedback

Commit:
:fire: [#234] Remove filter from update page

If we want to edit the list, we get all the zaken, so we need all the zaaktypes

SilviaAmAm commented on frontend/src/pages/destructionlist/detail/components/DestructionListEdit/DestructionListEdit.tsx:
```diff
@@ -134,6 +134,7 @@ export function DestructionListEdit() {
         ? selectableZaken
         : destructionListItems,
     isEditingState ? [...zakenOnPage, ...selectedUrls] : [],
+    undefined,
```
In the edit view, we get all the zaken, so we want all zaaktypes

Commit:
:green_heart: [#234] Fix E2E tests

Commit:
:green_heart: [#234] Attempt to fix storybook tests
